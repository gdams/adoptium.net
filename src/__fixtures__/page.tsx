import {
    AllReleaseNotes,
    AsciidocPage,
    File,
    ReleaseNotePage
} from '../types';

export const createAsciidocData = (): {
    asciidoc: AsciidocPage;
    file: File;
  } => ({
    asciidoc: {
        id: 'asciidoc-1',
        html: `
            <p>Hello World, this is a <pre><code>test</code></pre></p>
            <table><thead><td>Header 1</td><td>Header 2</td></thead><tbody><tr><td>Cell 1</td><td class="icon">Cell 2</td></tr></tbody></table>
            <a href="https://www.eclipse.org">Eclipse</a>
            <span class="icon">[docker]</span>
            <span class="icon">[check]</span>
        `,
        document: {
            title: 'Asciidoc Page title',
            main: 'Asciidoc Page content',
        },
        pageAttributes: {
            authors: 'author1, author2',
        },
        fields: {
            slug: '/asciidoc/asciidoc-page-title',
        }
    },
    file: {
        relativePath: 'test.adoc',
    }
});

export const createReleaseNoteData = (): {
    allReleaseNotes: AllReleaseNotes;
  } => ({
    allReleaseNotes: {
        edges: [
            {
                node: {
                    id: 'release-note-1',
                    version: '1.0.0',
                    title: 'Release Note title',
                    priority: 1,
                    component: 'component1',
                    link: 'https://release-note.com',
                }
            }
        ]
    }
});

