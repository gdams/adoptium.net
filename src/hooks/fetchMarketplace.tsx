const baseUrl = 'https://staging-marketplace-api.adoptium.net';
    
export async function getAllPkgsForVersion(
        version: number,
        os: string,
        architecture: string,
        package_type: string,
        checkboxRef: any,
    ): Promise<apiData> {
    let microsoftSelected = checkboxRef.current.vendorMicrosoft.checked;
    let temurinSelected = checkboxRef.current.vendorAdoptium.checked;
    let redhatSelected = checkboxRef.current.vendorRedHat.checked;
    let huaweiSelected = checkboxRef.current.vendorHuawei.checked;
    let zuluSelected = checkboxRef.current.vendorAzul.checked;
    let ibmSelected = checkboxRef.current.vendorIBM.checked;

    let params = '?'
    params += 'feature_version=' + version;

    if (os !== 'any' ) {
        if (os === 'alpine-linux') {
            params += '&os=alpine_linux'
        } else {
            params += ('&os=' + os)
        }
    }

    if (architecture !== 'any' ) {
        params += ('&architecture=' + architecture)
    }

    if (package_type === 'any') {
        params += '&image_type=jdk&image_type=jre'
    } else {
        params += ('&image_type=' + package_type)
    }

    if (temurinSelected) {
        params += '&vendor=adoptium'
    }

    if (redhatSelected) {
        params += '&vendor=redhat'
    }

    if (huaweiSelected) {
        params += '&vendor=huawei'
    }

    if (microsoftSelected) {
        params += ('&vendor=microsoft')
    }

    if (zuluSelected) {
        params += ('&vendor=azul')
    }

    if (ibmSelected) {
        params += ('&vendor=ibm')
    }

    let url = baseUrl + '/v1/assets/latestForVendors' + params;
    let json = await getPkgs(url);
    const data = JSON.parse(json);
    return data
}

async function getPkgs(url) {
    let response = await makeRequest("GET", url);
    return response;
}

async function makeRequest(method, url): Promise<apiData> {
    const response = await fetch(url);
    const apiResult = await response.text();
    return apiResult
};

export function getImageForDistribution(distribution: string) {
    switch(distribution) {
    case 'microsoft': return '/images/microsoft-logo.png';
    case 'temurin': return '/images/adoptium-logo.png';
    case 'redhat': return '/images/redhat.svg';
    case 'bisheng': return '/images/huawei.svg';
    case 'zulu': return '/images/azul-logo.png';
    case 'semeru_certified': return '/images/ibm-logo.png';
    default: return '';
    }
}
