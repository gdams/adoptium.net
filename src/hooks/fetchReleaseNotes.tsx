import { VersionMetaData } from '.';
import { useEffect, useState } from 'react';

const baseUrl = 'https://api.adoptium.net/v3/info/release_notes';

export function fetchReleaseNotesForVersion(
    isVisible: boolean,
    version: any,
): ReleaseNoteAPIResponse | null {
    if (!version) {
        return null
    }

    const [releaseNotes, setReleaseNotes] = useState<ReleaseNoteAPIResponse | null>(null);
    useEffect(() => {
        if (isVisible) {
        (async () => {
            setReleaseNotes(await fetchReleaseNote(version));
        })();
        }
    }, [isVisible]);

    return releaseNotes;
}

async function fetchReleaseNote(version) {
    const url = `${baseUrl}/${version}`;
    try {
        // fetch the data from the API
        const response = await fetch(url);
        return await response.json();
    } catch (error) {
        return { release_notes: null };
    }
}

export interface ReleaseNoteAPIResponse {
    id: string;
    release_name: string;
    release_notes: ReleaseNote[];
    vendor: string;
    version_data: VersionMetaData;
}

export interface ReleaseNote {
    id: string;
    link: URL;
    title: string;
    backportOf?: string;
    priority?: string;
    component?: string;
    subcomponent?: string;
    type?: string;
}
