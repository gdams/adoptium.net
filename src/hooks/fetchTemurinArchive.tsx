import moment from "moment";

import { fetchExtension } from '../util/fetchExtension';

const baseUrl = 'https://api.adoptium.net/v3';

let releases = []

export async function getAssetsForVersion(version: any, releaseType: any, numBuilds: any, buildDate: Date) {
  let url = `${baseUrl}/assets/feature_releases/${version}/${releaseType}?vendor=eclipse`;
  if (numBuilds) {
    url += `&page_size=${numBuilds}`
  }
  if (buildDate) {
    url += `&before=${moment(buildDate).format('Y-MM-DD')}`
  }
  releases = []
  let   json      = await makeRequest('GET', url);
  const response  = JSON.parse(json);
  const data      = response;
  let   pkgsFound = []
  for (let i = 0 ; i < data.length ; i++) {
      pkgsFound.push(data[i]);
  }
  return renderReleases(pkgsFound);
}

async function makeRequest(method: string, url: RequestInfo): Promise<apiData> {
    const response = await fetch(url);
    const apiResult = await response.text();
    return apiResult
};

function renderReleases(pkgs) {
  pkgs.forEach((aRelease) => {
    const release = {
        release_name: aRelease.release_name,
        openjdk_version: aRelease.version_data.openjdk_version,
        release_link: aRelease.release_link,
        timestamp: aRelease.timestamp,
        platforms: {},
    };

    aRelease.binaries.forEach((aReleaseAsset: { os: any; architecture: any; image_type: string; package: { link: any; checksum: any; size: number; }; installer: { link: any; checksum: any; size: number; }; }) => {
        const platform = `${aReleaseAsset.os}-${aReleaseAsset.architecture}`
  
        // Skip this asset if its platform could not be matched (see the website's 'config.json')
        if (!platform) {
          return;
        }
  
        // Skip this asset if it's not a binary type we're interested in displaying
        const binary_type = aReleaseAsset.image_type.toUpperCase();
        if (binary_type == 'SOURCES') {
          release.source_url = aReleaseAsset.package.link;
        }
        if (!['INSTALLER', 'JDK', 'JRE'].includes(binary_type)) {
          return;
        }
  
        if (!release.platforms[platform]) {
          release.platforms[platform] = {
            assets: [],
          };
        }
  
        let binary_constructor = {
          os: aReleaseAsset.os,
          architecture: aReleaseAsset.architecture,
          type: binary_type,
          link: aReleaseAsset.package.link,
          checksum: aReleaseAsset.package.checksum,
          size: Math.floor(aReleaseAsset.package.size / 1000 / 1000),
          extension: fetchExtension(aReleaseAsset.package.link)
        };
  
        if (aReleaseAsset.installer) {
          binary_constructor.installer_link = aReleaseAsset.installer.link;
          binary_constructor.installer_checksum = aReleaseAsset.installer.checksum;
          binary_constructor.installer_size =  Math.floor(aReleaseAsset.installer.size / 1000 / 1000);
          binary_constructor.installer_extension = fetchExtension(aReleaseAsset.installer.link)
        }
  
        // Add the new binary to the release asset
        release.platforms[platform].assets.push(binary_constructor);
    });
    releases.push(release);
  })
  return releases
}
