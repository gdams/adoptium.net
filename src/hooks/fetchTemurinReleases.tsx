import { fetchExtension } from '../util/fetchExtension';

const baseUrl = 'https://api.adoptium.net/v3';

export async function loadLatestAssets(version: number, os: string, architecture: string, packageType: string) {
    let url = `${baseUrl}/assets/latest/${version}/hotspot?`;
    if (os !== 'any') {
        url += `os=${os}&`
    }
    if (architecture !== 'any') {
        url += `architecture=${architecture}&`
    }
    if (packageType !== 'any') {
        url += `image_type=${packageType}&`
    }
    let   json      = await makeRequest('GET', url);
    const response  = JSON.parse(json);
    const data      = response;
    let   pkgsFound = []
    for (let i = 0 ; i < data.length ; i++) {
        pkgsFound.push(data[i]);
    }
    return renderReleases(pkgsFound);
}

async function makeRequest(method, url): Promise<apiData> {
    const response = await fetch(url);
    const apiResult = await response.text();
    return apiResult
};

function renderReleases(pkgs) {
    let releases = []
    pkgs.forEach((releaseAsset) => {
        const platform = `${releaseAsset.binary.os}-${releaseAsset.binary.architecture}`

        // Skip this asset if its platform could not be matched (see the website's 'config.json')
        if (!platform) {
            return;
        }

        // Skip this asset if it's not a binary type we're interested in displaying
        const binary_type = releaseAsset.binary.image_type.toUpperCase();
        if (binary_type == 'SOURCES') {
            releases['source'] = releaseAsset;
        }
        releases['releaseName'] = releaseAsset.version.openjdk_version;
        if (!['INSTALLER', 'JDK', 'JRE'].includes(binary_type)) {
            return;
        }
        // Get the existing release asset (passed to the template) or define a new one
        let release = releases.find((release) => release.platform_name === platform);
        if (!release) {
            release = {
                platform_name: `${releaseAsset.binary.os}-${releaseAsset.binary.architecture}`,
                os: releaseAsset.binary.os,
                architecture: releaseAsset.binary.architecture,
                release_name: releaseAsset.release_name,
                release_link: releaseAsset.release_link,
                release_date: releaseAsset.binary.updated_at,
                binaries: []
            };
        }

        let binary_constructor = {
            type: binary_type,
            link: releaseAsset.binary.package.link,
            checksum: releaseAsset.binary.package.checksum,
            size: Math.floor(releaseAsset.binary.package.size / 1000 / 1000),
            extension: fetchExtension(releaseAsset.binary.package.link)
        };

        if (releaseAsset.binary.installer) {
            binary_constructor.installer_link = releaseAsset.binary.installer.link;
            binary_constructor.installer_checksum = releaseAsset.binary.installer.checksum;
            binary_constructor.installer_size =  Math.floor(releaseAsset.binary.installer.size / 1000 / 1000);
            binary_constructor.installer_extension = fetchExtension(releaseAsset.binary.installer.link);
        }

        // Add the new binary to the release asset
        release.binaries.push(binary_constructor);

        // We have the first binary, so add the release asset.
        if (release.binaries.length === 1) {
            releases.push(release);
        }

        releases = orderPlatforms(releases, 'platform_ordinal');
        releases.forEach((release) => {
            release.binaries.sort((binaryA, binaryB) => binaryA.type > binaryB.type ? 1 : binaryA.type < binaryB.type ? -1 : 0);
        });
    })
    return releases
}

function orderPlatforms (input, attr = 'thisPlatformOrder') {
  return sortByProperty(input, attr, false);
};

function sortByProperty (input, property, descending) {
  const invert = descending ? -1 : 1;
  const sorter = (a, b) => {
    return invert * (a[property] > b[property] ? 1 : a[property] < b[property] ? -1 : 0);
  };

  if (Array.isArray(input)) {
    return input.sort(sorter);
  } else {
    // Preserve the source object key as '_key'
    return Object.keys(input)
      .map(_key => Object.assign(input[_key], {_key}))
      .sort(sorter);
  }
};
