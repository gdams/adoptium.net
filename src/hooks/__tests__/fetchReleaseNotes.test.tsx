import { renderHook, waitFor } from '@testing-library/react'
import { describe, expect, it, vi } from 'vitest'
import { fetchReleaseNotesForVersion } from '../fetchReleaseNotes';
import { createMockReleaseNotesAPI } from '../../__fixtures__/hooks';

let mockResponse = createMockReleaseNotesAPI(1);

global.fetch = vi.fn(() => Promise.resolve({
  json: () => Promise.resolve(mockResponse)
}));

afterEach(() => {
  vi.clearAllMocks();
});

describe('fetchReleaseNotesForVersion', () => {
  it('returns valid JSON', async () => {
    const { result } = renderHook(() => fetchReleaseNotesForVersion(true, 'sample_version'));
    await waitFor(() => {
      expect(result.current?.release_name).toBe('release_name_mock')
    }, { interval: 1 });
    expect(global.fetch).toHaveBeenCalledTimes(1)
    expect(fetch).toHaveBeenCalledWith(
      "https://api.adoptium.net/v3/info/release_notes/sample_version"
    );
    expect(result.current).toMatchSnapshot()
  })

  it('returns null if version is not defined', async () => {
    const { result } = renderHook(() => fetchReleaseNotesForVersion(true, null));
    expect(result.current).toBe(null)
  })
});