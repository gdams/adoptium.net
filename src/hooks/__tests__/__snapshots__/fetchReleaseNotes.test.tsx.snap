// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fetchReleaseNotesForVersion > returns valid JSON 1`] = `
{
  "id": "id_mock",
  "release_name": "release_name_mock",
  "release_notes": [
    {
      "id": "0",
      "link": "https://link_mock/",
      "priority": "1",
      "title": "title_mock",
    },
  ],
  "vendor": "vendor_mock",
  "version_data": {
    "build": 0,
    "major": 0,
    "minor": 0,
    "openjdk_version": "openjdk_version_mock",
    "patch": 0,
    "security": 0,
  },
}
`;
